src/hex_tarball.erl

```diff
@@ -0,0 +1,454 @@
+guess_build_tools(Metadata) ->
+    Metadata.
+
+%%====================================================================
+%% Tar Helpers
+%%====================================================================
+
+unpack_tarball(ContentsBinary, memory) ->
+    hex_erl_tar:extract({binary, ContentsBinary}, [memory, compressed]);
+unpack_tarball(ContentsBinary, Output) ->
+    hex_erl_tar:extract({binary, ContentsBinary}, [{cwd, Output}, compressed]).
+
+create_tarball(Files, Options) ->
+    Tarball = create_memory_tarball(Files),
+    case proplists:get_bool(compressed, Options) of
+        true -> gzip(Tarball);
+        false -> Tarball
+    end.
+
+create_memory_tarball(Files) ->
+    {ok, Fd} = file:open([], [ram, read, write, binary]),
+    {ok, Tar} = hex_erl_tar:init(Fd, write, fun file_op/2),
+
+    try
+        try
+            add_files(Tar, Files)
+        after
+            ok = hex_erl_tar:close(Tar)
+        end,
+
+        {ok, Size} = file:position(Fd, cur),
+        {ok, Binary} = file:pread(Fd, 0, Size),
+        Binary
+    after
+        ok = file:close(Fd)
+    end.
+
+file_op(write, {Fd, Data}) -> file:write(Fd, Data);
+file_op(position, {Fd, Pos}) -> file:position(Fd, Pos);
+file_op(read2, {Fd, Size}) -> file:read(Fd, Size);
+file_op(close, _Fd) -> ok.
+
+add_files(Tar, Files) when is_list(Files) ->
+    lists:map(fun(File) -> add_file(Tar, File) end, Files).
+
+add_file(Tar, {Filename, Contents}) when is_list(Filename) and is_binary(Contents) ->
+    ok = hex_erl_tar:add(Tar, Contents, Filename, tar_opts());
+add_file(Tar, Filename) when is_list(Filename) ->
+    {ok, FileInfo} = file:read_link_info(Filename, []),
+
+    case FileInfo#file_info.type of
+        symlink ->
+            ok = hex_erl_tar:add(Tar, Filename, tar_opts());
+        directory ->
+            ok = hex_erl_tar:add(Tar, Filename, tar_opts());
+        _ ->
+            Mode = FileInfo#file_info.mode,
+            {ok, Contents} = file:read_file(Filename),
+            ok = hex_erl_tar:add(Tar, Contents, Filename, Mode, tar_opts())
+    end.
+
+tar_opts() ->
+    NixEpoch = calendar:datetime_to_gregorian_seconds({{1970, 1, 1}, {0, 0, 0}}),
+    Y2kEpoch = calendar:datetime_to_gregorian_seconds({{2000, 1, 1}, {0, 0, 0}}),
+    Epoch = Y2kEpoch - NixEpoch,
+    [{atime, Epoch}, {mtime, Epoch}, {ctime, Epoch}, {uid, 0}, {gid, 0}].
+
+%% Reproducible gzip by not setting mtime and OS
+%%
+%% From https://tools.ietf.org/html/rfc1952
+%%
+%% +---+---+---+---+---+---+---+---+---+---+
+%% |ID1|ID2|CM |FLG|     MTIME     |XFL|OS | (more-->)
+%% +---+---+---+---+---+---+---+---+---+---+
+%%
+%% +=======================+
+%% |...compressed blocks...| (more-->)
+%% +=======================+
+%%
+%% +---+---+---+---+---+---+---+---+
+%% |     CRC32     |     ISIZE     |
+%% +---+---+---+---+---+---+---+---+
+gzip(Uncompressed) ->
+    Compressed = gzip_no_header(Uncompressed),
+    Header = <<31, 139, 8, 0, 0, 0, 0, 0, 0, 0>>,
+    Crc = erlang:crc32(Uncompressed),
```

Comment:
Just `Crc = crc32(Uncompressed)` works, right?

Reply:
no, most functions from `erlang` are auto-imported but not all and not this one (TIL)